Decode missing values in DataLoader and make regions respect existing missing values

spencerkclark opened the pull request:
Closes #190

spencerkclark committed:
Make regions respect existing invalid values

spencerahill commented on aospy/region.py:
```diff
@@ -166,7 +168,8 @@ def ts(self, data):
         sfc_area = data.sfc_area
         land_mask = _get_land_mask(data, self.do_land_mask)
-        weights = _sum_over_lat_lon((self.mask_var(sfc_area)*land_mask))
+        weights = _sum_over_lat_lon(
```
Could you add a comment that motivates the `np.isfinite` inclusion?

spencerahill commented on aospy/test/test_region.py:
```diff
@@ -0,0 +1,93 @@
+
+
+@pytest.fixture()
+def data():
```
Let's rename this to something more descriptive, e.g. `data_for_reg_calcs`

spencerahill commented on aospy/test/test_region.py:
```diff
@@ -0,0 +1,93 @@
+    )
+
+
+def test_mask_var(data, region_no_land_mask):
```
Can we add an analogous test for `region_land_mask` for completeness

spencerahill replied:
Or maybe incorporate it into this test via pytest's parameterize functionality

spencerahill commented on aospy/test/test_region.py:
```diff
@@ -0,0 +1,93 @@
+
+def test_ts_no_land_mask(data, region_no_land_mask):
+    result = region_no_land_mask.ts(data)
+    expected = xr.DataArray((0.25 * 4. + 0.5 * 3.) / (0.25 + 0.5))
```
Can you add a note explaining where these values come from?  And/or better yet pull them directly from the test data.

spencerahill reviewed (commented):
Thanks @spencerkclark for taking the time to do this, especially the tests.  A few minor things, mostly for clarification, but otherwise looks good.

spencerkclark committed:
Address review comments; fix mask_and_scale issue

spencerkclark commented:
@spencerahill thanks for the review comments; I went ahead and fixed the `mask_and_scale` issue as well (and added a test).  This should fully take care of #190 now.  Let me know if you have any more comments.

spencerkclark committed:
Add what's new entries

spencerahill commented on aospy/test/test_data_loader.py:
```diff
@@ -492,6 +492,24 @@ def preprocess(ds, **kwargs):
         np.testing.assert_allclose(result.values, expected.values)
 
+    def test_load_variable_mask_and_scale(self):
+        def preprocess(ds, **kwargs):
```
Name this `convert_all_to_missing_val` in place of the comment.

spencerahill commented on aospy/test/test_data_loader.py:
```diff
@@ -492,6 +492,24 @@ def preprocess(ds, **kwargs):
+            datetime(5, 12, 31),
+            intvl_in='monthly')
+
+        result = np.isfinite(result).sum().item()
```
Either rename this or the above `result` to a different name.  E.g. for this one `result_integral`.

spencerahill reviewed (commented):
Thanks @spencerkclark, nice work as usual.  Two final nits, then please go ahead and merge this. 👍

spencerkclark committed:
Rename variables in test_load_variable_mask_and_scale for clarity

spencerkclark commented:
Thanks @spencerahill, merging now.
